_stbt/frameobject.py

```diff
@@ -413,3 +413,56 @@ def is_visible(self):
         raise NotImplementedError(
             "Objects deriving from FrameObject must define an is_visible "
             "property")
+
+    def refresh(self, frame=None, **kwargs):
+        r"""Returns a new FrameObject typically of the same type as `self`, but
+        with un updated frame.  `self` is not modified.
+
+        If you want to override this behaviour implement `_refresh` on your
+        derived class instead.
+
+        .. testsetup::
+
+            >>> import stbt
+            >>> from tests.test_frame_object import _load_frame
+            >>> class Dialog(stbt.FrameObject):
+            ...     @property
+            ...     def is_visible(self):
+            ...         return bool(self._info)
+            ...
+            ...     @property
+            ...     def title(self):
+            ...         return stbt.ocr(region=stbt.Region(396, 249, 500, 50),
+            ...                         frame=self._frame)
+            ...
+            ...     @property
+            ...     def message(self):
+            ...         right_of_info = stbt.Region(
+            ...             x=self._info.region.right, y=self._info.region.y,
+            ...             width=390, height=self._info.region.height)
+            ...         return stbt.ocr(region=right_of_info, frame=self._frame) \
+            ...                .replace('\n', ' ')
+            ...
+            ...     @property
+            ...     def _info(self):
+            ...         return stbt.match('tests/info.png', frame=self._frame)
+
+        >>> page = Dialog(frame=_load_frame('with-dialog'))
+        >>> print page.message
+        This set-top box is great
+        >>> page = page.refresh(_load_frame('with-dialog2'))
+        >>> print page.message
+        This set-top box is fabulous
+
+        Added in v30"""
+        if frame is None:
+            import stbt
+            frame = stbt.get_frame()
+        return self._refresh(frame=frame, **kwargs)
+
+    @classmethod
+    def _refresh(cls, frame):
+        """Override this method on derived classes to customise refresh
+        behaviour.  Unlike with `refresh` the `frame` parameter is guaranteed
+        to be a `numpy.ndarray` and not be `None`."""
+        return cls(frame=frame)
```

Comment:
I wonder if the "non-None frame" guarantee is useful enough to warrant having 2 different methods `refresh` and `_refresh`.

This could just as easily be:

```
@classmethod
def refresh(cls frame=None):
    return cls(frame=frame)
```

The FrameObject constructor already takes care of getting a new frame if `frame` is None.

Did you have any other (future) functionality in mind for `refresh` that you don't want subclasses to have to re-implement themselves?

Reply:
> I wonder if the "non-None frame" guarantee is useful enough to warrant having 2 different methods refresh and _refresh.

Yeah, it's a bit old-school OOP.  One motivation was both to cut down on boilerplate.  I think I'd find myself writing:

    if frame is None:
        import stbt
        frame = stbt.get_frame()

quite a lot which for simple functions can be longer than the rest of the function.

> Did you have any other (future) functionality in mind for refresh that you don't want subclasses to have to re-implement themselves?

Yes. I was considering using other techniques than `stbt.get_frame()` to retrieve the new frame.  For example we could keep a `dut` member from which to pull frames which might help to write tests that involve multiple STBs at the same time.